### ISSUE TITLE
SNDH tracks don't stop playing

### ISSUE THREAD

**Issue by chris-y** [frno7/psgplay#21]
In the SNDH archive, file `Daglish_Ben/Footballer_of_the_Year_2.sndh` plays a couple of seconds then the rest is silent.
In interactive mode it shows that it is still playing track 1 although there is no sound.
Is it possible to get psgplay to move onto the next track when the first one has finished playing?  `--stop=auto` isn't working for this particular file.
Maybe silence detection is required?

_Originally mentioned in https://github.com/frno7/psgplay/issues/13#issuecomment-1321181917_

**Comment by frno7** [comment]
The `TIME` tag in the SNDH format is a provision for the length of a track in seconds:

https://github.com/frno7/psgplay/blob/d719a9bf353fef637cb759d49039a9a80624bc57/doc/sndhv21.txt#L53-L54

Sadly, most SNDH files don’t have it. Most proper tracks, that aren’t sound effects, programmatically plays for ever in a loop, and so silence is an unreliable means of track length determination. Many tracks that _do_ have `TIME` have it at `0`, meaning looping for ever anyway.

I’ve seen efforts to measure true SNDH track lengths. Maybe someone maintaining the [SNDH archive](http://sndh.atari.org/) could be convinced it would be worthwhile for an update? Best would be millisecond resolution, to avoid spurious bleeps and hisses at track endings!

The `--info` option displays all SNDH tags. Here’s one exceptional example having a `TIME` of 208 seconds:

```sh
% psgplay --info Mad_Max/Last_Ninja.sndh
path Mad_Max/Last_Ninja.sndh
tag field TITL Last Ninja
tag field COMM Mad Max
tag field RIPP Grazey / PHF
tag field CONV Grazey / PHF
tag field FLAG ~ y PSG
tag field TIME 1 208
header size 94
data size 10476
```

Since `TIME` is so rarely available, PSG play doesn’t really make good use of it, as of yet.

**Comment by benjihan** [comment]
Hi,

sc68 has a database for sndh track length. Lengths are automatically measured using mksc68 time command. I am supposed to update the whole sndh database someday with the duration and FLAG tag updated. Meanwhile the database is hardcoded [here](https://sourceforge.net/p/sc68/code/HEAD/tree/file68/src/timedb.inc.h).
I can also provide the latest database to this day if you are interested. It looks something like that:
 
`sndh YsaA -------------------------------- fffaeaae 01 03:27 00010362 00050 00-69bdfd3c19e36780eda12f7d18ac3f24 02477 - No More - leif Rullar conversion.sndh`

--
Ben/OVR

**Comment by frno7** [comment]
> `TIMEDB_ENTRY( fffaeaae, 1, 10362, TB+TD+YM), /* 02477 - No More - leif Rullar conversion */`

Hmm, the `10362` entry must be... half the resolution of a [_centisecond_](https://en.wiktionary.org/wiki/centisecond)? Which is to say the count of 1/50 seconds, so the vertical blank timing of PAL? :-) Much better than full seconds.

> I am supposed to update the whole sndh database someday with the duration and FLAG tag updated.

Nice! How are things proceeding?

**Comment by benjihan** [comment]
> Hmm, the `10362` entry must be... half the resolution of a [_centisecond_](https://en.wiktionary.org/wiki/centisecond)? Which is to say the count of 1/50 seconds, so the vertical blank timing of PAL? :-) Much better than full seconds.

The length is the number of frames to run. Hence the duration depends on this value and the replay rate. Like you said most of the time it's a 50hz PAL VBL per frame. A hash code is used to identify the (unpacked) file. The algorithm  can be found in the [file68.c::isread()](https://sourceforge.net/p/sc68/code/HEAD/tree/file68/src/file68.c#l414) function.

> Nice! How are things proceeding?

Unfortunately I've been slacking on this for quiet sometime now. The thing is I have a completely rewritten version of sc68 in progress more like stalled right now. 

I've just updated the [timedb.inc.h](https://sourceforge.net/p/sc68/code/HEAD/tree/file68/src/timedb.inc.hl) file.

**Comment by benjihan** [comment]
> Hmm, the `10362` entry must be... half the resolution of a [_centisecond_](https://en.wiktionary.org/wiki/centisecond)? Which is to say the count of 1/50 seconds, so the vertical blank timing of PAL? :-) Much better than full seconds.

The length is the number of frames to run. Hence the duration depends on this value and the replay rate. Like you said most of the time it's a 50hz PAL VBL per frame. A hash code is used to identify the (unpacked) file. The algorithm  can be found in the [file68.c::isread()](https://sourceforge.net/p/sc68/code/HEAD/tree/file68/src/file68.c#l414) function.

> Nice! How are things proceeding?

Unfortunately I've been slacking on this for quiet sometime now. The thing is I have a completely rewritten version of sc68 in progress more like stalled right now. 

I've just updated the [timedb.inc.h](https://sourceforge.net/p/sc68/code/HEAD/tree/file68/src/timedb.inc.hl) file.

**Comment by frno7** [comment]
> The length is the number of frames to run. Hence the duration depends on this value and the replay rate. Like you said most of the time it's a 50hz PAL VBL per frame.

Is it feasible to add a millisecond resolution (or there about) SNDH tag, that is simple to interpret as track length, for SNDH players?

> Unfortunately I've been slacking on this for quiet sometime now. The thing is I have a completely rewritten version of sc68 in progress more like stalled right now.

Oh, how did the desire to rewrite sc68 come about?

> I've just updated the [timedb.inc.h](https://sourceforge.net/p/sc68/code/HEAD/tree/file68/src/timedb.inc.h) file.
> ```
> TIMEDB_ENTRY( 503196f8,  1,        84,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  2,       132,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  3,       517,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  4,        64,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  5,        27,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  6,        56,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  7,        42,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  8,        42,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8,  9,        82,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8, 10,        67,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> TIMEDB_ENTRY( 503196f8, 11,      4952,      YM), /* 00851 - Ben Daglish - Footballer of the Year 2 */
> ```

As suspected, only track 11 is anywhere near a length of a proper tune. :-)

**Comment by benjihan** [comment]
> Is it feasible to add a millisecond resolution (or there about) SNDH tag, that is simple to interpret as track length, for SNDH players?

It's certainly possible. I don't think Grazey or Evil would mind.The number of frames is the most accurate you can get. sc68 files store the first run length and the loop length (0=non looping). I understand having to do mul/div on somewhat big numbers can be an hassle in 68k/asm  `ms = (uint64_t) frames * 1000 / replay_rate`. May be a { uint8_t  h, m, s, centi ) or a bcd coded $HMMSSMSE.
IIRC the current sndh TIME tag can not be used for multiple song files.    

> Oh, how did the desire to rewrite sc68 come about?

Mainly the STe DMA/LMC  simulation really needed work. I was not happy with the m68k simulator compiled code being huge because of generated code. That was fast at the time but useless nowaday. The project need an overall  cleaning.

> As suspected, only track 11 is anywhere near a length of a proper tune. :-)

Indeed. One thing I wanted to add with the `FLAG` tag is a `JINGLE` flag and a `SFX` flag. Duration works fairly well as a proxy.

**Comment by frno7** [comment]
> IIRC the current sndh TIME tag can not be used for multiple song files.

`Mad_Max/Games/SID/Warp.sndh` has a `TIME` for every track it has, no?

```
% psgplay --info Mad_Max/Games/SID/Warp.sndh
path Mad_Max/Games/SID/Warp.sndh
tag field TITL Warp - Sid
tag field COMM Mad Max
tag field RIPP Grazey / PHF
tag field CONV Grazey / PHF
tag field YEAR 2017
tag field ## 8
tag field !# 1
tag field !V 50
tag field TIME 1 219
tag field TIME 2 80
tag field TIME 3 70
tag field TIME 4 95
tag field TIME 5 148
tag field TIME 6 20
tag field TIME 7 6
tag field TIME 8 93
header size 126
data size 12388
```

**Comment by benjihan** [comment]
> `Mad_Max/Games/SID/Warp.sndh` has a `TIME` for every track it has, no?

Right. I was not sure about that. I remember now `TIME` argument is an array of word. I thought it was stored as a string  I was confusing it with `##`

**Comment by frno7** [comment]
@chris-y, with commit e6da8e7ebd9dfb99c578344d1a50769172487d54, the length, if available as a tag in the SNDH, is indicated next to each track. It doesn’t proceed to the next track automatically, yet, but it seems to be a good idea to have it do that. Example:

```
PSG play                           01:23
Mad Max
Warp - Sid

> 1. Warp - Sid                    03:39
  2. Warp - Sid                    01:20
  3. Warp - Sid                    01:10
  4. Warp - Sid                    01:35
  5. Warp - Sid                    02:28
  6. Warp - Sid                    00:20
  7. Warp - Sid                    00:06
  8. Warp - Sid                    01:33
```

**Comment by chris-y** [comment]
On the command line it should stop at the correct time too?

**Comment by frno7** [comment]
Should _command mode_ stop and quit after a track, whereas _interactive mode_ proceeds to the subsequent track? After all, as a command it can easily be scripted, to do other things.

**Comment by chris-y** [comment]
Yes, I would suggest so. I suppose it could be an option to play through.
Maybe if a track is specified, it stops after playing, and if not it starts at track 1 and plays through them all?

**Comment by frno7** [comment]
The SNDH format has a default track

https://github.com/frno7/psgplay/blob/e6da8e7ebd9dfb99c578344d1a50769172487d54/doc/sndhv21.txt#L47

and so, unless a track is specified, the command may as well play the last track, if it happens to be the default. Of course, the command could still run through them all, if it plays track 1 after the last track, and then stops at the track preceding the default track. Someone will have to write a good user’s manual for this, I fear... :-)

**Comment by frno7** [comment]
There’s also a slight complication in the fact that tracks may be expressly _endless_

https://github.com/frno7/psgplay/blob/e6da8e7ebd9dfb99c578344d1a50769172487d54/doc/sndhv21.txt#L118-L119

meaning it won’t stop, quit or proceed automatically. Perhaps a `--length-default=<[mm:]ss[.ss]>` option could terminate tracks that are either indeterminate or expressly endless. Tracks that do have finite lengths would still play to completion, regardless.

**Comment by chris-y** [comment]
Shouldn't --length=xxxx do this already?

**Comment by frno7** [comment]
> Shouldn't --length=xxxx do this already?

The `--length` option overrides lengths for tracks that do have finite lengths, meaning that if SNDH specifies that a certain track is, say, 3 minutes, `--length=02:00` would stop it after 2 minutes. The `--length-default` option would only override tracks that don’t have a length specified by SNDH or are expressly endless.

**Comment by frno7** [comment]
> Is it possible to get psgplay to move onto the next track when the first one has finished playing?

@chris-y, this has now been implemented, and is also default.

Tunes are required to have the new `FRMS` (tune duration) tag for this to work. Most tunes in the [2026 archive](https://sndh.atari.org/download.php) have it (with some exceptions).

New [keyboard controls](https://github.com/frno7/psgplay#interactive-text-mode) in the past few days:

- `r` or `left arrow` to rewind 1 second (during pause or play);
- `f` or `right arrow` to fast forward 1 second (during pause or play);
- `shift`-`R` to toggle repeat tunes;
- `shift`-`S` to toggle single or all tunes.

Are your ready to close this issue? :-)
